pkg/system/phase2_creating.go

```diff
@@ -171,6 +176,128 @@ func (r *Reconciler) ReconcilePhaseCreatingForMainClusters() error {
 	return nil
 }
 
+func (r *Reconciler) UpgradeDbPostgres16() error {
+	var err error = nil
+
+	if r.NooBaa.Status.PostgresUpdatePhase == nbv1.UpgradePhaseFinished {
+		r.Logger.Infof("UpgradePostgresDB: DB is already upgraded to postgresql-16")
+		return nil
+	}
+
+	switch r.NooBaa.Status.PostgresUpdatePhase {
+	case "":
+		// Check whether it is fresh install case
+		sts := util.KubeObject(bundle.File_deploy_internal_statefulset_postgres_db_yaml).(*appsv1.StatefulSet)
+		sts.Name = "noobaa-db-pg"
+		sts.Namespace = r.NooBaa.Namespace
+		stsObj, err := util.KubeGetObject(sts)
+		// Postgres database doesn't exists, no need to upgrade.
+		if err != nil {
+			r.Logger.Infof("UpgradePostgresDB: old STS doesn't exist - no need for upgrade")
+			r.NooBaa.Status.PostgresUpdatePhase = nbv1.UpgradePhaseNone
+			r.NooBaa.Status.PostgresMajorVersion = nbv1.PostgresMajorVersionV16
+		} else {
+			desiredImage := GetDesiredDBImage(r.NooBaa, "")
+			// Check whether upgrade is reqd, set phase depending the requirement
+			if r.isPostgresUpgradeReqd(desiredImage, stsObj.(*appsv1.StatefulSet)) {
+				r.Logger.Infof("UpgradePostgresDB: upgrading postgres16")
+				r.NooBaa.Status.PostgresUpdatePhase = nbv1.UpgradePhasePrepare
+			} else {
+				if r.NooBaa.Status.PostgresMajorVersion == nbv1.PostgresMajorVersionV16 {
+					r.NooBaa.Status.PostgresMajorVersion = nbv1.PostgresMajorVersionV16
+					r.NooBaa.Status.PostgresUpdatePhase = nbv1.UpgradePhaseFinished
+				} else {
+					r.NooBaa.Status.PostgresUpdatePhase = nbv1.UpgradePhaseNone
+				}
+			}
+		}
+	case nbv1.UpgradePhaseNone:
+		// Check whether it is fresh install case
+		sts := util.KubeObject(bundle.File_deploy_internal_statefulset_postgres_db_yaml).(*appsv1.StatefulSet)
+		sts.Name = "noobaa-db-pg"
+		sts.Namespace = r.NooBaa.Namespace
+		obj, err := util.KubeGetObject(sts)
+		if err != nil {
+			r.Logger.Infof("UpgradePostgresDB: STS doesn't exist - no need for upgrade")
+			return err
+		}
+		desiredImage := GetDesiredDBImage(r.NooBaa, "")
+		// Check wheter upgrade is reqd, set phase depending the requirement
+		if r.isPostgresUpgradeReqd(desiredImage, obj.(*appsv1.StatefulSet)) {
+			r.Logger.Infof("UpgradePostgresDB: upgrading postgres16")
+			r.NooBaa.Status.PostgresUpdatePhase = nbv1.UpgradePhasePrepare
+		} else {
+			if r.NooBaa.Status.PostgresMajorVersion == nbv1.PostgresMajorVersionV16 {
+				r.NooBaa.Status.PostgresMajorVersion = nbv1.PostgresMajorVersionV16
+				r.NooBaa.Status.PostgresUpdatePhase = nbv1.UpgradePhaseFinished
+			} else {
+				r.NooBaa.Status.PostgresUpdatePhase = nbv1.UpgradePhaseNone
+			}
+		}
+	case nbv1.UpgradePhasePrepare:
+		if err := r.setEndpointsDeploymentReplicas(0); err != nil {
+			r.Logger.Errorf("UpgradePostgresDB::got error on endpoints deployment reconcile %v", err)
+			return err
+		}
+		desiredImage := GetDesiredDBImage(r.NooBaa, "")
```

Comment:
Yes. sure. It was kept outside earlier then I thouht to minimize the scope just to avoid calling it for unnecessary cases. Now we are checking "finished" start of the function itself, its okay to move it next to it